streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/TaskAssignorConvergenceTest.java

```diff
@@ -358,6 +369,57 @@ private static void runRandomizedScenario(final long seed) {
         }
     }
 
+    private static void verifyBalancedAssignment(final Harness harness, final int balanceFactor) {
+        final int activeDiff;
+        final int activeStatefulDiff;
+        final double activeStatelessPerThreadDiff;
+        final int assignedDiff;
+        final int standbyDiff;
+
+        {
+            final int maxActive = harness.clientStates.values().stream().map(ClientState::activeTaskCount).max(comparingInt(i -> i)).orElse(0);
+            final int minActive = harness.clientStates.values().stream().map(ClientState::activeTaskCount).min(comparingInt(i -> i)).orElse(0);
+            activeDiff = maxActive - minActive;
+        }
+        {
+            final int maxActiveStateful = harness.clientStates.values().stream().map(s -> diff(TreeSet::new, s.activeTasks(), harness.statelessTasks).size()).max(comparingInt(i -> i)).orElse(0);
+            final int minActiveStateful = harness.clientStates.values().stream().map(s -> diff(TreeSet::new, s.activeTasks(), harness.statelessTasks).size()).min(comparingInt(i -> i)).orElse(0);
+            activeStatefulDiff = maxActiveStateful - minActiveStateful;
+        }
+        {
+            final double maxActiveStatefulPerThread = harness.clientStates.values().stream().map(s1 -> 1.0 * intersection(TreeSet::new, s1.activeTasks(), harness.statelessTasks).size() / s1.capacity()).max(comparingDouble(i -> i)).orElse(0.0);
```

Comment:
The HATA originally tried to balance each type of task individually (ie stateful active, standby, stateless active) and IIRC you made a convincing argument against doing that during the review and for balancing only the total task load. What's the rationale for enforcing this now? Or did I misremember and/or misinterpret your earlier point

Reply:
FWIW we'll probably want to revisit this if/when we migrate standby processing to a separate thread(s) but I'm +1 for enforcing this now. Just wondering what the current reasoning is.

On the other hand, maybe it's better to enforce a good data parallelism over an equal per-task-type balance. Like if subtopology 1 is significantly heavier than subtopology 2, then it could be more balanced to have active 1_0 and active 2_0 on one instance and standby 1_0 and standby 2_0 on the other. But maybe that's not just worth optimizing 
 at this point

Reply:
I can neither confirm nor deny any statements that I may or may not have made in the past ;)  Kidding aside, I don't recall what that convincing argument might have been. Perhaps just that that kind of multivariate optimization would be complicated to implement, and maybe it's better to just start simple.

Whatever the intent might have been, though, the actual implementation of our balancer should already achieve a "perfect" balance of stateful-active tasks. I'm not sure about standbys, but I don't see why we shouldn't just make them balanced as well.

Regarding stateless tasks, it does seem suboptimal to assign them to instances that already have active tasks when there are instances available with no tasks at all.